### PR TITLE
[8403008] Add client connection related edgeHub metrics

### PR DESCRIPTION
**Pull request by davilu**
1. Added following metrics

- (Gauge) `edgehub_connected_clients`: current number of connected clients to edge hub
- (Counter) `edgehub_client_connect_success_total`: total number of times individual client successfully connect to edgeHub
- (Counter) `edgehub_client_disconnect_total`: total number of times individual client disconnected from edgeHub

2. fixed (Counter) `edgehub_client_connect_failed_total` report failure while re-auth failed

### CODE REVIEW

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceConnectionMetrics.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        static readonly List<string> EmptyStringList = new List<string>();
+        static readonly string[] EmptyStringArray = new string[0];
+        public static readonly IMetricsGauge ConnectedClientsGauge = EdgeMetrics.Instance.CreateGauge(
+            "connected_clients",
+            "Current number of connected clients to edge hub",
+            EmptyStringList);
+
+        public static readonly IMetricsCounter ClientsConnectCounter = EdgeMetrics.Instance.CreateCounter(
+            "client_connect_success",
+            "Total number of times individual client successfully connect to edgeHub",
```
Comment (veyalla) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceConnectionMetrics.cs:19]:
```suggestion
            "Total number of times each client successfully connected to edgeHub",
```

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceConnectionMetrics.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            new List<string>() { "id" });
+
+        public static readonly IMetricsCounter ClientsDiscconnectCounter = EdgeMetrics.Instance.CreateCounter(
+           "client_disconnect",
+           "Total number of times individual client disconnected from edgeHub",
```
Comment (veyalla) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceConnectionMetrics.cs:24]:
```suggestion
           "Total number of times each client disconnected from edgeHub",
```

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceConnectionMetrics.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    using Microsoft.Azure.Devices.Edge.Util.Metrics;
+    using EdgeMetrics = Util.Metrics.Metrics;
+
+    public static class DeviceConnectionMetrics
+    {
+        static readonly List<string> EmptyStringList = new List<string>();
+        static readonly string[] EmptyStringArray = new string[0];
+        public static readonly IMetricsGauge ConnectedClientsGauge = EdgeMetrics.Instance.CreateGauge(
+            "connected_clients",
+            "Current number of connected clients to edge hub",
```
Comment (veyalla) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceConnectionMetrics.cs:14]:
```suggestion
            "Current number of clients connected to edge hub",
```

File: doc/BuiltInMetrics.md [hunk 4]
```diff
@@ -21,25 +21,27 @@ instance_number | A Guid representing the current runtime. On restart, all metri
...
+| `edgehub_offline_duration_seconds`                          |  `id` (Module ID)<br> `quantile`(Percentile [50, 90, 95, 99, 99.9, 99.99]) 									| P50, P90, P95, P99, P99.9 and P99.99 time edge hub was offline. Values may be reported as `NaN`  if no new measurements are reported for a  certain period of time (currently 10 minutes). As this is `summary` type, corresponding `_count` and `_sum` counters will be emitted.              	| summary |
+| `edgehub_operation_retry_total`      						  |  `id` (Module ID)<br>`operation` (Operation name)                                                        	| Total number of times edgeHub operations were retried 					| counter | 
+| `edgehub_client_connect_failed_total`                       | `id` (Device/Module ID)<br> `reason` (not authenticated)<br> 												| Total number of times individual client failed to connect to edgeHub		| counter |                                                              
+| `edgehub_connected_clients`                              	  | 															 												| Current number of connected clients to edge hub        					| gauge |
```
Comment (veyalla) [inline doc/BuiltInMetrics.md:42]:
```suggestion
| `edgehub_connected_clients`                              	  | 															 												| Current number of clients connected to edgeHub        					| gauge |                                                            
```

File: doc/BuiltInMetrics.md [hunk 5]
```diff
@@ -21,25 +21,27 @@ instance_number | A Guid representing the current runtime. On restart, all metri
...
+| `edgehub_operation_retry_total`      						  |  `id` (Module ID)<br>`operation` (Operation name)                                                        	| Total number of times edgeHub operations were retried 					| counter | 
+| `edgehub_client_connect_failed_total`                       | `id` (Device/Module ID)<br> `reason` (not authenticated)<br> 												| Total number of times individual client failed to connect to edgeHub		| counter |                                                              
+| `edgehub_connected_clients`                              	  | 															 												| Current number of connected clients to edge hub        					| gauge |                                                            
+| `edgehub_client_connect_success_total`                      | `id` (Device/Module ID)<br>																					| Total number of times individual client successfully connect to edgeHub   | counter |                                                                                                                                                                    | counter |
```
Comment (veyalla) [inline doc/BuiltInMetrics.md:43]:
```suggestion
| `edgehub_client_connect_success_total`                      | `id` (Device/Module ID)<br>																					| Total number of times each client successfully connected to edgeHub   | counter |                                                                                                                                                                    | counter |                                                              
```

File: doc/BuiltInMetrics.md [hunk 6]
```diff
@@ -21,25 +21,27 @@ instance_number | A Guid representing the current runtime. On restart, all metri
...
+| `edgehub_client_connect_failed_total`                       | `id` (Device/Module ID)<br> `reason` (not authenticated)<br> 												| Total number of times individual client failed to connect to edgeHub		| counter |                                                              
+| `edgehub_connected_clients`                              	  | 															 												| Current number of connected clients to edge hub        					| gauge |                                                            
+| `edgehub_client_connect_success_total`                      | `id` (Device/Module ID)<br>																					| Total number of times individual client successfully connect to edgeHub   | counter |                                                                                                                                                                    | counter |                                                              
+| `edgehub_client_disconnect_total`                      	  | `id` (Device/Module ID)<br>																					| Total number of times individual client disconnected from edgeHub   		| counter |
```
Comment (veyalla) [inline doc/BuiltInMetrics.md:44]:
```suggestion
| `edgehub_client_disconnect_total`                      	  | `id` (Device/Module ID)<br>																					| Total number of times each client disconnected from edgeHub   		| counter |
```

File: doc/BuiltInMetrics.md [hunk 7]
```diff
@@ -21,25 +21,27 @@ instance_number | A Guid representing the current runtime. On restart, all metri
...
+| `edgehub_offline_count_total`                               |  `id` (Module ID)<br>                                                      									| Total number of times edgeHub went offline								| counter | 
+| `edgehub_offline_duration_seconds`                          |  `id` (Module ID)<br> `quantile`(Percentile [50, 90, 95, 99, 99.9, 99.99]) 									| P50, P90, P95, P99, P99.9 and P99.99 time edge hub was offline. Values may be reported as `NaN`  if no new measurements are reported for a  certain period of time (currently 10 minutes). As this is `summary` type, corresponding `_count` and `_sum` counters will be emitted.              	| summary |
+| `edgehub_operation_retry_total`      						  |  `id` (Module ID)<br>`operation` (Operation name)                                                        	| Total number of times edgeHub operations were retried 					| counter | 
+| `edgehub_client_connect_failed_total`                       | `id` (Device/Module ID)<br> `reason` (not authenticated)<br> 												| Total number of times individual client failed to connect to edgeHub		| counter |
```
Comment (veyalla) [inline doc/BuiltInMetrics.md:41]:
```suggestion
| `edgehub_client_connect_failed_total`                       | `id` (Device/Module ID)<br> `reason` (not authenticated)<br> 												| Total number of times each client failed to connect to edgeHub		| counter |                                                              
```

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceConnectionMetrics.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+namespace Microsoft.Azure.Devices.Edge.Hub.Core
+{
+    using System.Collections.Generic;
+    using Microsoft.Azure.Devices.Edge.Util.Metrics;
+    using EdgeMetrics = Util.Metrics.Metrics;
+
+    public static class DeviceConnectionMetrics
+    {
+        static readonly List<string> EmptyStringList = new List<string>();
+        static readonly string[] EmptyStringArray = new string[0];
```
Comment (yophilav) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceConnectionMetrics.cs:11]: please consider using `Array.Empty<string>()`

File: doc/BuiltInMetrics.md [hunk 9]
```diff
@@ -21,25 +21,27 @@ instance_number | A Guid representing the current runtime. On restart, all metri
...
+| `edgehub_operation_retry_total`      						  |  `id` (Module ID)<br>`operation` (Operation name)                                                        	| Total number of times edgeHub operations were retried 					| counter | 
+| `edgehub_client_connect_failed_total`                       | `id` (Device/Module ID)<br> `reason` (not authenticated)<br> 												| Total number of times each client failed to connect to edgeHub		| counter |                                                              
+| `edgehub_connected_clients`                              	  | 															 												| Current number of clients connected to edgeHub        					| gauge |                                                            
+| `edgehub_client_connect_success_total`                      | `id` (Device/Module ID)<br>																					| Total number of times each client successfully connected to edgeHub   | counter |                                                                                                                                                                    | counter |
```
Comment (yophilav) [inline doc/BuiltInMetrics.md:43]: 1. With the current implementation, there is no distinction between have the same module connected & reconnected multiple times vs having multiple modules connect at the same time. I'm not sure what the customer use case is, but I'm having a hard time interpreting the meaning of this metric.
2. This metric only accurate per edgeHub restart cycle.

Comment (davilu) [reply doc/BuiltInMetrics.md:43]: 1. There is `id` in the metrics which will tell which device/module it is. The counter will group based on the `id`.
2. Ideally metrics should be collected periodically. In user scenario if customer  would like to monitor edge hub, the metrics should be feed to Azure monitoring or tracking or ELK search.  Restart causes metrics data rest shouldn't be considered in the implementation unless we'd like to persist it locally.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceConnectionMetrics.cs (hunk 8 above)
Comment (davilu) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceConnectionMetrics.cs:11]: Will do. There is a interesting article [here](https://www.tabsoverspaces.com/233833-best-way-to-create-an-empty-collection-array-and-list-in-csharp-net) about performance.

**Review by lfitchett — APPROVED** [review]
LGTM

**Pull request merged**